pallets/api/src/messaging/transports/ismp.rs

```diff
@@ -164,90 +156,120 @@ impl<T: Config> IsmpModule for Handler<T> {
+	// Static as not in use.
 	fn on_accept(&self, _request: &PostRequest) -> Weight {
-		todo!()
+		DbWeightOf::<T>::get().reads_writes(1, 1)
 	}
 
-	fn on_timeout(&self, _timeout: &Timeout) -> Weight {
-		DbWeightOf::<T>::get().reads_writes(2, 1)
+	fn on_timeout(&self, timeout: &Timeout) -> Weight {
+		let x = match timeout {
+			Timeout::Request(Request::Post(_)) => 0u32,
+			Timeout::Request(Request::Get(_)) => 1u32,
+			Timeout::Response(_) => 2u32,
+		};
+		T::WeightInfo::ismp_on_timeout(x)
 	}
 
-	fn on_response(&self, _response: &Response) -> Weight {
-		DbWeightOf::<T>::get().reads_writes(2, 2)
-	}
-}
-
-fn calculate_deposit<T: Config>(mut deposit: BalanceOf<T>) -> BalanceOf<T> {
-	// Add amount for `IsmpRequests` lookup.
-	let key_len: BalanceOf<T> =
-		(T::AccountId::max_encoded_len() + MessageId::max_encoded_len()).saturated_into();
-	deposit.saturating_accrue(
-		T::ByteFee::get() * (H256::max_encoded_len().saturated_into::<BalanceOf<T>>() + key_len),
-	);
+	// todo: test
+	fn on_response(&self, response: &Response) -> Weight {
+		let x = match response {
+			Response::Get(_) => 1,
+			Response::Post(_) => 0,
+		};
 
-	deposit
+		T::WeightInfo::ismp_on_response(x).saturating_add(T::CallbackExecutor::execution_weight())
+		// Also add actual weight consumed by contract env.
+	}
 }
 
-fn process_response<T: Config>(
+pub(crate) fn process_response<T: Config>(
 	commitment: &H256,
-	encode: &impl Encode,
-	store: impl Fn() -> Vec<u8>,
+	response_data: &impl Encode,
 	event: impl Fn(AccountIdOf<T>, MessageId) -> Event<T>,
 ) -> Result<(), anyhow::Error> {
-	let (origin, id) =
-		IsmpRequests::<T>::get(commitment).ok_or(Error::Custom("request not found".into()))?;
+	ensure!(
+		response_data.encoded_size() <= T::MaxResponseLen::get() as usize,
+		Error::Custom("Response length exceeds maximum allowed length.".into())
+	);
 
-	let Some(super::super::Message::Ismp { commitment, callback, deposit }) =
-		Messages::<T>::get(&origin, &id)
+	let (initiating_origin, id) =
+		IsmpRequests::<T>::get(commitment).ok_or(Error::Custom("Request not found.".into()))?;
+
+	let Some(super::super::Message::Ismp { commitment, callback, message_deposit }) =
+		Messages::<T>::get(&initiating_origin, id)
 	else {
-		return Err(Error::Custom("message not found".into()))?
+		return Err(Error::Custom("Message must be an ismp request.".into()).into());
 	};
 
+	// Deposit that the message has been recieved before a potential callback execution.
+	Pallet::<T>::deposit_event(event(initiating_origin.clone(), id));
+
 	// Attempt callback with result if specified.
 	if let Some(callback) = callback {
-		// TODO: check response length
-		if Pallet::<T>::call(origin.clone(), callback, id, &encode, deposit).is_ok() {
-			Pallet::<T>::deposit_event(event(origin, id));
+		if Pallet::<T>::call(&initiating_origin, callback, &id, response_data).is_ok() {
+			// Clean storage, return deposit
+			Messages::<T>::remove(&initiating_origin, id);
```

Comment:
Shared be a shared remove function to ensure that all cleanup is done consistently across the pallet, otherwise maintainers need to know to update cleanup handling in x different places.